File: src/Api/Woof.Api/DataAccess/YamlFileStore.cs
using Woof.Api.DataAccess.Models;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace Woof.Api.DataAccess;

public class YamlFileStore<T, U> : IFileStore<T>
    where T : FileEntity
    where U : IStep
{
    private readonly string _filePath;
    private List<Action<List<T>>> _commands = new();
    
    private ISerializer Serializer = new SerializerBuilder()
        .WithNamingConvention(CamelCaseNamingConvention.Instance)
        .Build();

    private IDeserializer Deserializer;

    private static SemaphoreSlim Gate = new(1);

    private YamlFileStore(string filePath, IDictionary<string, Type> typeDiscriminators)
    {
        _filePath = filePath;
        Deserializer = new DeserializerBuilder()
        .WithNamingConvention(CamelCaseNamingConvention.Instance)
        .WithTypeDiscriminatingNodeDeserializer(o =>
        {
            o.AddKeyValueTypeDiscriminator<U>("type", typeDiscriminators);
        })
        .Build();
    }

    public static YamlFileStore<T, U> Create(IWebHostEnvironment env, string fileName, IDictionary<string, Type> typeDiscriminators)
    {
        var filePath = Path.Combine(env.WebRootPath, fileName);

        if (!File.Exists(filePath))
            File.WriteAllText(filePath, "[]");

        return new YamlFileStore<T, U>(filePath, typeDiscriminators);
    }

    public async Task<TResult> QueryAsync<TResult>(Func<List<T>, TResult> query)
    {
        await Gate.WaitAsync();
        var text = await File.ReadAllTextAsync(_filePath);
        var data = Deserializer.Deserialize<List<T>>(text);

        var result = query(data);
        Gate.Release();

        return result;
    }

    public void Command(Action<List<T>> command) => _commands.Add(command);

    public void Update(T entity)
    {
        _commands.Add(xs => _commands.Add(xs => UpdateCommand(xs, entity)));
    }

    public Task UpdateAsync(T entity)
    {
        _commands.Add(xs => UpdateCommand(xs, entity));
        return CompleteAsync();
    }

    private void UpdateCommand(List<T> xs, T target)
    {
        var i = xs.FindIndex(x => x.Id == target.Id);
        if (i == -1)
            throw new InvalidOperationException($"Entity {target.Id} not found");

        xs[i] = target;
    }

    public async Task CompleteAsync()
    {
        await Gate.WaitAsync();
        var text = await File.ReadAllTextAsync(_filePath);
        var data = Deserializer.Deserialize<List<T>>(text);

        foreach (var cmd in _commands)
            cmd(data);

        _commands.Clear();
        var newText = Serializer.Serialize(data);
        await File.WriteAllTextAsync(_filePath, newText);
        Gate.Release();
    }
}

public static class YamlFileStoreExtensions
{
    public static void AddYamlFileStore<T, U>(
        this IServiceCollection services,
        IWebHostEnvironment env,
        string fileName, IDictionary<string, Type> typeDiscriminators)
        where T : FileEntity
        where U : IStep
    {
        services.AddScoped<IFileStore<T>>(_ => YamlFileStore<T, U>.Create(env, fileName, typeDiscriminators));
    }
}

File: src/Api/Woof.Webapi/DataAccess/YamlFileStore.fs
namespace Woof.Webapi.DataAccess

open System.IO
open YamlDotNet.Serialization
open YamlDotNet.Serialization.NamingConventions
open System.Threading
open Microsoft.AspNetCore.Hosting

type YamlFileStore<'T> private (filePath: string) =
    let filePath = filePath

    let mutable commands: list<list<'T> -> list<'T>> = [];
    
    static member serializer =
        SerializerBuilder().WithNamingConvention(CamelCaseNamingConvention.Instance).Build()

    static member deserializer =
        DeserializerBuilder().WithNamingConvention(CamelCaseNamingConvention.Instance).Build()

    static member semaphore = new SemaphoreSlim(1)

    static member Create<'T>(env: IWebHostEnvironment, fileName: string) =
        let filePath = Path.Combine(env.WebRootPath, fileName)
        match File.Exists(filePath) with
        | true -> ()
        | false -> File.WriteAllText(filePath, "[]");    

        YamlFileStore<'T>(filePath)

    member this.FindAsync(predicate: 'T -> bool) = task {
        do! YamlFileStore<'T>.semaphore.WaitAsync()
        let! text = File.ReadAllTextAsync(filePath)
        let data = YamlFileStore<'T>.deserializer.Deserialize<list<'T>>(text)
        do YamlFileStore<'T>.semaphore.Release() |> ignore

        let result =
            data
            |> List.find predicate

        return result
    }

    member this.Do(command: list<'T> -> list<'T>) = commands <- command::commands
        
    member this.Complete() = task {
        do! YamlFileStore<'T>.semaphore.WaitAsync()
        let! text = File.ReadAllTextAsync(filePath)
        let mutable data = YamlFileStore<'T>.deserializer.Deserialize<list<'T>>(text)

        for cmd in commands do
            data <- cmd data

        let yaml = YamlFileStore<'T>.serializer.Serialize(data)
        do! File.WriteAllTextAsync(filePath, yaml)
        commands <- []
    }